### PR TITLE
Support serving mTLS and user mTLS/TLS on same port

### LINKED ISSUE #51768: Dual ingress sidecar TLS/istio mTLS termination on the same port
**Describe the feature request**
* Currently, when supporting internal traffic using istio mTLS and external traffic using simple or mutual tls, two ports must be defined and a PeerAuthentication config disabling istio mTLS on the external TLS port must be defined (see [ingress sidecar termination](https://istio.io/latest/docs/tasks/traffic-management/ingress/ingress-sidecar-tls-termination/) example). I would like to have the ability to configure the same port to serve both traffic.
* In the original [RFC](https://docs.google.com/document/d/15Qhr7errbylXEzxxCK7ij_oUpn4E5SFU2uDdl_n2GIc/edit#heading=h.xw1gqgyqs5b), supporting the same port was not considered due to it being unclear how to distinguish Istio mTLS from external TLS. However, since Istio mTLS uses magic ALPN protocols, I suggest we distinguish them using ALPN listener matches
* I propose either defining a boolean variable in the Sidecar resource (e.g. `hybridPort: true`) to distinguish this from existing use cases, or merging IstioIngressListener configurations if it is determined that there is no conflict between two that share the same port. 
* Alternatively, we can handle this if the authorizationpolicy is [permissive](https://github.com/istio/istio/blob/cd8829f703dd4d5d4a3dfed85a32310ea7fd2ab1/pilot/pkg/networking/core/listener_inbound.go#L220-L231)
* Would also need to distinguish the protocol after termination

**Describe alternatives you've considered**
* Rewrite services that send traffic to the external TLS port to send traffic to the internal Istio mTLS port

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[ ] Dual Stack
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

### PR DESCRIPTION
**Pull request by wtzhang23**
**Please provide a description of this PR:**
Allow serving mTLS with Istio ALPNs and user mTLS/TLS on the same port if PeerAuthentication is permissive. See [RFC](https://docs.google.com/document/d/13ciYV5H85rOc_EH1LcuoVyCFxFu_n4SpNq1luvHmRmA/edit#heading=h.xw1gqgyqs5b).

Fixes #51768

### CODE REVIEW

**Comment by wtzhang23** [comment]
Tested via the following script
```sh
TEST_DIR=/tmp/test-istio-dual-tls
mkdir -p ${TEST_DIR}
GOBIN=${TEST_DIR} go install istio.io/release-builder@latest

envsubst <<EOF > ${TEST_DIR}/build-manifest.yaml
version: 1.24.0
docker: docker.io/istio
directory: ${TEST_DIR}/istio-release
dependencies:
  istio:
    git: https://github.com/wtzhang23/istio
    branch: dual-tls-ingress
outputs: ["docker", "helm"]
skipGenerateBillOfMaterials: true
EOF
${TEST_DIR}/release-builder build --manifest ${TEST_DIR}/build-manifest.yaml

kind create cluster --name istio-test-dual-tls --kubeconfig ${TEST_DIR}/.kubeconfig
for archive in ${TEST_DIR}/istio-release/out/docker/*; do
    echo "Preloading ${archive}"
    kind load image-archive --name istio-test-dual-tls <(gzip --decompress --stdout ${archive})
done
kubectl create namespace --kubeconfig ${TEST_DIR}/.kubeconfig istio-system
helm install istio-base ${TEST_DIR}/istio-release/out/helm/base-1.24.0.tgz -n istio-system --kubeconfig ${TEST_DIR}/.kubeconfig
helm install istiod ${TEST_DIR}/istio-release/out/helm/istiod-1.24.0.tgz -n istio-system --kubeconfig ${TEST_DIR}/.kubeconfig --set env.ENABLE_TLS_ON_SIDECAR_INGRESS=true --wait

kubectl create namespace --kubeconfig ${TEST_DIR}/.kubeconfig test-server
kubectl label namespace --kubeconfig ${TEST_DIR}/.kubeconfig test-server istio-injection=enabled --overwrite
openssl req -x509 -newkey rsa:4096 -sha256 -days 3650 -nodes -keyout ${TEST_DIR}/tls.key -out ${TEST_DIR}/tls.crt -subj "/CN=foo.bar"
kubectl create secret tls --kubeconfig ${TEST_DIR}/.kubeconfig -n test-server tls-cert --cert ${TEST_DIR}/tls.crt --key ${TEST_DIR}/tls.key
cat <<EOF | kubectl apply --kubeconfig ${TEST_DIR}/.kubeconfig -f -
apiVersion: security.istio.io/v1
kind: PeerAuthentication
metadata:
  name: default
  namespace: test-server
spec:
  mtls:
    mode: PERMISSIVE
---
apiVersion: networking.istio.io/v1
kind: Sidecar
metadata:
  name: test-tls
  namespace: test-server
spec:
  workloadSelector:
    labels:
      app: test-server
  ingress:
    - port:
        number: 80
        protocol: HTTPS
      tls:
        mode: SIMPLE
        serverCertificate: /etc/tls-secrets/tls.crt
        privateKey: /etc/tls-secrets/tls.key
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-server
  namespace: test-server
  labels:
    app: test-server
spec:
  selector:
    matchLabels:
      app: test-server
  template:
    metadata:
      labels:
        app: test-server
      annotations:
        sidecar.istio.io/userVolumeMount: '[{"name":"tls-cert","mountPath":"/etc/tls-secrets","readOnly":true}]'
    spec:
      containers:
      - name: httpbin
        image: kennethreitz/httpbin
        ports:
        - containerPort: 80
      volumes:
      - name: tls-cert
        secret:
          secretName: tls-cert
---
apiVersion: v1
kind: Service
metadata:
  name: test-server
  namespace: test-server
spec:
  selector:
    app: test-server
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
EOF


kubectl create namespace --kubeconfig ${TEST_DIR}/.kubeconfig test-client-mesh
kubectl label namespace --kubeconfig ${TEST_DIR}/.kubeconfig test-client-mesh istio-injection=enabled --overwrite
cat <<EOF | kubectl apply --kubeconfig ${TEST_DIR}/.kubeconfig -f -
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-client-mesh
  namespace: test-client-mesh
  labels:
    app: test-client-mesh
spec:
  selector:
    matchLabels:
      app: test-client-mesh
  template:
    metadata:
      labels:
        app: test-client-mesh
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
      volumes:
      - name: tls-cert
        secret:
          secretName: tls-cert
EOF

kubectl create namespace --kubeconfig ${TEST_DIR}/.kubeconfig test-client
cat <<EOF | kubectl apply --kubeconfig ${TEST_DIR}/.kubeconfig -f -
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-client
  namespace: test-client
  labels:
    app: test-client
spec:
  selector:
    matchLabels:
      app: test-client
  template:
    metadata:
      labels:
        app: test-client
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
      volumes:
      - name: tls-cert
        secret:
          secretName: tls-cert
EOF

sleep 40

kubectl exec --kubeconfig ${TEST_DIR}/.kubeconfig -n test-server deploy/test-server -c istio-proxy -- curl localhost:15000/config_dump > ${TEST_DIR}/config_dump.json
kubectl exec --kubeconfig ${TEST_DIR}/.kubeconfig -n test-client-mesh deploy/test-client-mesh -- curl http://test-server.test-server.svc.cluster.local:80/headers
kubectl exec --kubeconfig ${TEST_DIR}/.kubeconfig -n test-client deploy/test-client -- curl --insecure https://test-server.test-server.svc:80/headers
```

**Comment by bleggett** [comment]
> **Please provide a description of this PR:** Allow serving Istio mTLS and user TLS on the same port if PeerAuthentication is permissive. See [RFC](https://docs.google.com/document/d/13ciYV5H85rOc_EH1LcuoVyCFxFu_n4SpNq1luvHmRmA/edit#heading=h.xw1gqgyqs5b).
> 
> Fixes #51768 Depends on #53396

I'm not sure exactly what you mean by `Istio mTLS` here - there is really no such thing.

Are you saying you want the same port to terminate TLS _and_ mTLS traffic, at the same time?

**Comment by wtzhang23** [comment]
Yes, sorry for my misleading naming. When I mention `Istio mTLS`, I really mean "mTLS using Istio-specific ALPNs". I'll update the wording.

**Comment by hzxuzhonghu** [comment]
To speed up the procedure, It is better to present on the meeting https://docs.google.com/document/d/1wsa06GGiq1LEGwhkiPP0FKIZJqdAiue-VeBonWAzAyk/edit#heading=h.o8pz6aqnzzgk

**Comment by wtzhang23** [comment]
Added doc to the list of those in review

File: pilot/pkg/networking/core/listener_inbound.go [hunk 1]
```diff
@@ -225,21 +236,36 @@ func (lb *ListenerBuilder) buildInboundListeners() []*listener.Listener {
...
 			lp := istionetworking.ModelProtocolToListenerProtocol(cc.port.Protocol)
 			opts = getTLSFilterChainMatchOptions(lp)
 			mtls.TCP = BuildListenerTLSContext(cc.tlsSettings, lb.node, lb.push.Mesh, istionetworking.TransportProtocolTCP, false)
 			mtls.HTTP = mtls.TCP
-		} else {
+			chains = append(chains, lb.inboundChainForOpts(cc, mtls, opts)...)
+		}
+
+		if cc.tlsSettings == nil || mtls.Mode != model.MTLSDisable {
 			lp := istionetworking.ModelProtocolToListenerProtocol(cc.port.Protocol)
```
Comment (wtzhang23) [inline pilot/pkg/networking/core/listener_inbound.go:265]: TODO: Will need to ensure that the permissive TLS chains are not built when `tlsSettings != nil`.

Comment (wtzhang23) [reply pilot/pkg/networking/core/listener_inbound.go:265]: Addressed in [8d29fdd](https://github.com/istio/istio/pull/53469/commits/8d29fdd1aaa3b6f1150ab32555306b693db6ec23)

**Review by howardjohn — CHANGES_REQUESTED** [review]
Looks pretty solid, just a few comments. Nice work!

File: pilot/pkg/networking/core/listener_inbound.go [hunk 2]
```diff
@@ -225,21 +236,36 @@ func (lb *ListenerBuilder) buildInboundListeners() []*listener.Listener {
...
+			mtls := mtls
+
 			// Since we are terminating TLS, we need to treat the protocol as if its terminated.
 			// Example: user specifies protocol=HTTPS and user TLS, we will use HTTP
 			cc.port.Protocol = cc.port.Protocol.AfterTLSTermination()
+
+			// In permissive mode, both the Istio mTLS filter chains and user TLS filter chains
+			// are built. Set the flag to ensure when constructing them they are uniquely referencable.
+			if mtls.Mode == model.MTLSPermissive {
+				cc.permissive = true
```
[the remark this answers is not in the document]
Comment (wtzhang23) [reply pilot/pkg/networking/core/listener_inbound.go:255]: Do you think the distinction between listener names is needed? Could keep it simple and use the same names for both.

**Comment by wtzhang23** [comment]
Any way to mark this as not stale?

**Pull request closed without merging**